# New features
- Export biochemical properties of gene regions with `-biochemicalProperty <geneFeature> <property>` or `-baseBiochemicalProperties <geneFeature>` export options. Available in export for alignments, clones and SHM tree nodes. Available properties: Hydropathy, Charge, Polarity, Volume, Strength, MjEnergy, Kf1, Kf2, Kf3, Kf4, Kf5, Kf6, Kf7, Kf8, Kf9, Kf10, Rim, Surface, Turn, Alpha, Beta, Core, Disorder, N2Strength, N2Hydrophobicity, N2Volume, N2Surface.

# Algorithm enhancement
- Global consensus assembly algorithm used in `assemble` to collapse UMI/Cell groups into contigs now have better seed selection empiric for multi-consensus assembly scenarios. This increases sensitivity during assembly of secondary consensuses from the same group of sequences.
- Consensus assembly parameters `maxNormalizedAlignmentPenalty` and `altSeedPenaltyTolerance` are adjusted to increase sensitivity. 

# Minor fixes
- Corrected the help message that appears when using a deprecated preset and incorrectly suggests using `--assemble-contigs-by` instead of `--assemble-clonotypes-by`.
- The `--split-by-sample` option is now set to `true` by default for all `align` presets, as well as all presets that inherit from it. This new default behavior applies unless it is directly overridden in the preset or with `--dont-split-by-sample` mix-in.
- Fixed possible crash with `--dry-run` option in `analyze`
- Fixed assembling feature for `irepertoire-human-rna-xcr-repseq-plus` preset. Now `{CDR2Begin:FR4End}`.

# New Presets
- `bd-sc-xcr-rhapsody-full-length-enhanced-bead-v2` new preset for BD full-length protocol with enhanced beads V2 featuring B384 whitelists.
- `takara-mouse-rna-tcr-umi-smarseq` preset for Takara Bio SMART-Seq Mouse TCR (with UMIs).
