# Presets Changes
- The `milab-human-rna-tcr-umi-race` preset has been updated. Now, clones are assembled by default based on the CDR3, in line with the manufacturer's recommended read length.
- The `flairr-seq-bcr` preset has been updated. Now, the preset sets species to `human` by default according to a built-in tag pattern with primer sequences.
- The following presets have been added to cover Ivivoscribe assay panels: `invivoscribe-human-dna-trg-lymphotrack`,`invivoscribe-human-dna-trb-lymphotrack`, `invivoscribe-human-dna-igk-lymphotrack`,`invivoscribe-human-dna-ighv-leader-lymphotrack`,`invivoscribe-human-dna-igh-fr3-lymphotrack`, `invivoscribe-human-dna-igh-fr2-lymphotrack`,`invivoscribe-human-dna-igh-fr1-lymphotrack`,`invivoscribe-human-dna-igh-fr123-lymphotrack`.
- The following ppresets have been added for mouse Thermofisher assays: `thermofisher-mouse-rna-tcb-ampliseq-sr`,`thermofisher-mouse-dna-tcb-ampliseq-sr`,`thermofisher-mouse-rna-igh-ampliseq-sr`,`thermofisher-mouse-dna-igh-ampliseq-sr`.

## üöÄ New features

### New characteristics in clonotype export

- `-aaLength` and `-allAALength` is available alongside `-nLength` and `-allNLength`
- `-aaMutationRate` is available alongside `-nMutationRate`
- Added optional arg `germline` in `-nFeature`, `-aaFeature`, `-nLength`, `-aaLength`
  in `exportClones`, `exportAlignments` and `exportCloneGroups`. It allows to export a sequence of the germline instead
  of a sequence of the gene.
- For all mutation exports (excluding -mutationsDetailed) added optional filter by mutation
  type: `[(substitutions|indels|inserts|deletions)]`
- Added `-nMutationsCount`, `-aaMutationsCount`, `-allNMutationsCount`, `-allAAMutationsCount` for all relatable exports
- For mutation exports in `exportShmTreesWithNodes` `(germline|mrca|parent)` option is now optional. Will be export
  mutations from `germline` by default
- Added `--export-clone-groups-sort-chains-by` mixin

### Deprecations of export options

- `-lengthOf` now is deprecated, use `-nLength` instead
- `-allLengthOf` now is deprecated, use `-allNLength` instead
- `-mutationRate` now is deprecated, use `-nMutationRate` instead

## üõ†Ô∏è Minor improvements & fixes

- Forbid export of germline sequence for VJJunction in shmTrees exports
- Fix validation of parameter in `-nMutationRate`
- Fix `-nMutationRate` if region is not covered for the clone
- Output of `exportAlignmentsPretty` in 4.5.0 it changed to verbose variant by default. Now we restore previous
  behaviour
- Fix IllegalArgumentException in `exportAlignmentsPretty` for cases where could not possible to translate nucleotide
  sequence to amino acids
- Fix error if `analyze` executed with `-f` and `--output-not-used-reads` in the same time
- Resolutions of wildcards are excluded from calculation of `-nMutationRate` for CDR3 in `exportShmTreesWithNodes`
- Fix OutOfMemory exception in command `extend` with .vdjca input
- In `findShmTrees` filter for productive only clones now check for stop codons in all features, not only in CDR3
- Change default value for filter for productive clones in `findShmTrees` to false (was true before)
- Add option `--productive-only` to `findShmTrees`
- Fix parsing of `--export-clone-groups-for-cell-type` parameter
- Fix usage of `slice` command on clnx files that weren't ordered by id.
- In `slice` now default behaviour is to keep original ids. Previous behaviour can be called by `--reassign-ids`
