## üöÄ New features and major changes

- Consensus assembly in `assemble` now is performed separately for each chain. This allows to prevent effects from
  different expression levels on the consensus assembly algorithm. This change is specifically important for single-cell
  presets with cell-level assembly (most of the MiXCR presets for single-cell data).

## üõ†Ô∏è Minor improvements & fixes

- Fixed parsing of optional arguments
  for `exportShmTreesWithNodes`: `-nMutationsRelative`, `-aaMutations`, `-nMutations`, `-aaMutationsRelative`, `-allNMutations`, `-allAAMutations`, `-allNMutationsCount`, `-allAAMutationsCount`.
- Fixed parsing of optional arguments for `exportClones`
  and `exportAlignments`: `-allNMutations`, `-allAAMutations`, `-allNMutationsCount`, `-allAAMutationsCount`.
- Fixed possible errors on exporting amino acid mutations in `exportShmTreesWithNodes`
- Fixed list of required options in `listPresets` command
- Fixed error on building trees in case of JBeginTrimmed started before CDR3Begin
- Fixed usage `--remove-step qc`
- Added `--remove-qc-check` option
- Remove `-topChains` field from `exportShmTreesWithNodes` command. Use `-chains` instead
- Removed default splitting clones by V and J for presets where clones are assembled by full-length.
- Fixed NullPointerException in some cases of building trees by SC+bulk data
- Fixed `java.lang.IllegalArgumentException: While adding VEndTrimmed` in `exportClones`
- Fixed combination trees step in `findShmTrees`: in some cases trees weren't combined even if it could be
- Fixed `java.util.NoSuchElementException` in some cases of SC combining of trees
- Fixed export of `-jBestIdentityPercent` in `exportShmTreesWithNodes`
- Added validation on export `-aaFeature` for features containing UTR
- Fixed usage of command `exportPlots shmTrees`
- Fixed topology of trees: before common V and J mutations were included in the root node, now in root included only
  reconstructed NDN. Previous behavior lead to smaller numbers of distance from germline, sequence for germline exported
  with common mutations. For fix this you should recalculate `findShmTrees`
- Fixed `IllegalStateException` on removing unnecessary genes on `findAlleles`
- Added `--dont-remove-unused-genes` option to `findAlleles` command
- Command `groupClones` was renamed to `assembleCells`. Old name is working, but it's hidden from help
