## ‚ùó Breaking changes
- Starting from version 4.7.0 of MiXCR, users are required to specify the assembling feature for all presets in cases where
it's not defined by the protocol. This can be accomplished using either the option ```--assemble-clones-by Feature``` 
or ```--assemble-contigs-by Feature``` for fragmented data (such as RNA-seq or 10x VDJ data). This ensures consistency in 
assembling features when integrating various samples or types of samples, such as 10x single-cell VDJ and AIRR sequencing data,
for downstream analyses like inferring alleles or building SHM trees. The previous behavior for fragmented data, which aimed 
to assemble as long sequences as possible, can still be achieved with either the option ```--assemble-contigs-by-cell``` 
for single-cell data or ```--assemble-longest-contigs``` for RNA-seq/Exom-seq data.

## üöÄ Major fixes and upgrades
- Ability to trigger realignments of left or right reads boundaries with global alignment algorythm using
parameters ```rightForceRealignmentTrigger Feature``` or ```leftForceRealignmentTrigger Feature``` in case the reads do 
not span the CDR3 regions (rescue alignments in case of fragmented single cell data).
- Fixed ```assemble``` behavior in presets for single-cell data (in some cases consensuses were assembled from reads coming
from different cells)
- Ability to override the ```relativeMeanScore``` and ```maxHits``` parameters in ```assemble``` and ```assembleContigs``` steps
(improve the V genes assignments)